apps/heft/src/pluginFramework/HeftTask.ts

```diff
@@ -0,0 +1,216 @@
+
+  public get parentPhase(): HeftPhase {
+    return this._parentPhase;
+  }
+
+  public get taskName(): string {
+    return this._taskName;
+  }
+
+  public get consumingTasks(): ReadonlySet<HeftTask> {
+    if (!this._consumingTasks) {
+      // Force initialize all dependency relationships
+      // This needs to operate on every phase in the set because the relationships are only specified
+      // in the consuming phase.
+      const { tasks } = this._parentPhase;
+
+      for (const task of tasks) {
+        task._consumingTasks = new Set();
+      }
+      for (const task of tasks) {
+        for (const dependency of task.dependencyTasks) {
+          dependency._consumingTasks!.add(task);
+        }
+      }
+    }
+
+    return this._consumingTasks!;
+  }
+
+  public get pluginDefinition(): HeftTaskPluginDefinition {
+    if (!this._taskPluginDefinition) {
+      throw new InternalError(
+        'HeftTask.ensureInitializedAsync() must be called before accessing HeftTask.pluginDefinition.'
+      );
+    }
+    return this._taskPluginDefinition;
+  }
+
+  public get pluginOptions(): object | undefined {
+    return this._taskSpecifier.taskEvent?.options || this._taskSpecifier.taskPlugin?.options;
+  }
+
+  public get dependencyTasks(): Set<HeftTask> {
+    if (!this._dependencyTasks) {
+      this._dependencyTasks = new Set();
+      const dependencyNamesSet: Set<string> = new Set(this._taskSpecifier.taskDependencies || []);
+
+      for (const dependencyName of dependencyNamesSet) {
+        // Skip if we can't find the dependency
+        const dependencyTask: HeftTask | undefined = this._parentPhase.tasksByName.get(dependencyName);
+        if (!dependencyTask) {
+          throw new Error(
+            `Could not find dependency task "${dependencyName}" within phase "${this._parentPhase.phaseName}"`
+          );
+        }
+        this._dependencyTasks.add(dependencyTask);
+      }
+    }
+
+    return this._dependencyTasks!;
+  }
+
+  public constructor(
+    parentPhase: HeftPhase,
+    taskName: string,
+    taskSpecifier: IHeftConfigurationJsonTaskSpecifier
+  ) {
+    this._parentPhase = parentPhase;
+    this._taskName = taskName;
+    this._taskSpecifier = taskSpecifier;
+
+    this._validate();
+  }
+
+  public async ensureInitializedAsync(): Promise<void> {
+    if (!this._taskPluginDefinition) {
+      this._taskPluginDefinition = await this._loadTaskPluginDefintionAsync();
+      this.pluginDefinition.validateOptions(this.pluginOptions);
+    }
+  }
+
+  public async getPluginAsync(logger: IScopedLogger): Promise<IHeftTaskPlugin<object | void>> {
+    await this.ensureInitializedAsync();
+    if (!this._taskPlugin) {
+      this._taskPlugin = await this._taskPluginDefinition!.loadPluginAsync(logger);
+    }
+    return this._taskPlugin;
+  }
+
+  private async _loadTaskPluginDefintionAsync(): Promise<HeftTaskPluginDefinition> {
+    if (this._taskSpecifier.taskEvent) {
```

Comment:
Can you add a defensive check to make sure both `taskEvent` and `taskPlugin` aren't specified?